File: backend/utils/pricing.py
from data.knowledge_base import get_knowledge_base

SUBSCRIBER_DEFAULT_DISCOUNT = 15


def apply_discount(price, discount_percent):
    discount = round((price * discount_percent) / 100)
    return max(0, price - discount)


def get_discount_for_item(offers, user_role, service_id=None):
    if user_role != "subscriber":
        return 0
    applicable = [
        o
        for o in offers
        if o.get("active", True)
        and (o.get("serviceId") is None or o.get("serviceId") == service_id)
    ]
    from_offers = max((o.get("discountPercent", 0) for o in applicable), default=0)
    return max(SUBSCRIBER_DEFAULT_DISCOUNT, from_offers)


def price_for_item(base_price, user_role, offers, service_id=None):
    price = float(base_price or 0)
    if user_role == "admin":
        return {"base": price, "final": price, "discountPercent": 0}
    if user_role == "subscriber":
        discount = get_discount_for_item(offers, user_role, service_id)
        return {
            "base": price,
            "final": apply_discount(price, discount),
            "discountPercent": discount,
        }
    return {"base": price, "final": price, "discountPercent": 0}


def merge_catalog(custom_services, user_role, offers):
    base = list(get_knowledge_base())
    for c in custom_services:
        base.append(
            {
                "id": c["id"],
                "title": c["title"],
                "category": c["category"],
                "price": c["basePrice"],
                "description": c.get("description", ""),
                "custom": True,
            }
        )
    result = []
    for s in base:
        p = price_for_item(s["price"], user_role, offers, s["id"])
        result.append(
            {
                **s,
                "basePrice": p["base"],
                "price": p["final"],
                "discountPercent": p["discountPercent"],
            }
        )
    return result

File: backend/utils/pricing.js
const knowledgeBase = require("../data/knowledgeBase");

const SUBSCRIBER_DEFAULT_DISCOUNT = 15;

function applyDiscount(price, discountPercent) {
  const discount = Math.round((price * discountPercent) / 100);
  return Math.max(0, price - discount);
}

function getOffersForUser(offers, userRole, serviceId = null) {
  if (userRole !== "subscriber") return [];
  return offers.filter(
    (o) =>
      o.active !== false &&
      (o.role === "subscriber" || !o.role) &&
      (o.serviceId == null || o.serviceId === serviceId)
  );
}

function priceForItem(basePrice, userRole, offers, serviceId = null) {
  const price = Number(basePrice) || 0;
  if (userRole === "admin") return { base: price, final: price, discountPercent: 0 };

  if (userRole === "subscriber") {
    const applicable = getOffersForUser(offers, "subscriber", serviceId);
    const best = applicable.reduce((max, o) => Math.max(max, o.discountPercent || 0), SUBSCRIBER_DEFAULT_DISCOUNT);
    return { base: price, final: applyDiscount(price, best), discountPercent: best };
  }

  return { base: price, final: price, discountPercent: 0 };
}

function enrichServices(services, userRole, offers) {
  return services.map((s) => {
    const pricing = priceForItem(s.price, userRole, offers, s.id);
    return {
      ...s,
      basePrice: pricing.base,
      price: pricing.final,
      discountPercent: pricing.discountPercent,
    };
  });
}

function mergeCatalog(customServices, userRole, offers) {
  const base = [...knowledgeBase, ...customServices.map((c) => ({
    id: c.id,
    title: c.title,
    category: c.category,
    price: c.basePrice,
    description: c.description,
    custom: true,
  }))];
  return enrichServices(base, userRole, offers);
}

module.exports = {
  SUBSCRIBER_DEFAULT_DISCOUNT,
  priceForItem,
  enrichServices,
  mergeCatalog,
  getOffersForUser,
};
